File: ruby/leetcode/majority-element-ii.rb
# @param {Integer[]} nums
# @return {Integer[]}
def majority_element(nums)
    seen = {}
    majority_count = nums.size / 3
    nums.each do |n|
        seen[n] = if seen[n] then seen[n] + 1 else 1 end
    end
    return seen.select { |k, v| v > majority_count }.keys
end

File: javascript/leetcode/majority-element-ii.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var majorityElement = function(nums) {
    const seen = {};
    if (nums.length === 1) return nums;
    const majorityCount = Math.floor(nums.length / 3);
    nums.forEach((n) => {
        seen[n] = seen[n] ? seen[n] + 1 : 1; 
    });
    return Object.keys(seen).filter((k) => seen[k] > majorityCount);
};
